Push unit tests down to document specs

benilovj opened the pull request:
Supersedes https://github.com/alphagov/specialist-publisher-rebuild/pull/759

The document model specs have a lot of duplication between them; however the models themselves (except `DrugSafetyUpdate`) don't actually contain the logic under test – the `Document` class does.

This PR moves the specs into `document_spec` and slims the specs down to a single shared spec. The shared spec checks that saving sends an appropriate payload to Publishing API, which is valid against the content schema. This step isn't possible to test in `document_spec`, because there isn't a schema to validate against for the dummy document subclass used in document specs.

Out of scope here (will look to do these on subsequent PRs):
- converting the `Payloads` module to use FactoryGirl
- tidying up `drug_safety_update_spec`

benilovj committed:
Add a Document spec for successful publishing

This is inspired by the successful publishing spec for the document
subclasses.

benilovj committed:
Create a factory for Publishing API payloads

This factory generates a plausible content item, and can be extended
for specific specialist document formats

benilovj committed:
Remove redundant #publish! specs for specialist document types

These cases are all covered by `document_spec`.

benilovj committed:
Add spec for Document#save!

benilovj committed:
Add a Document spec for unsuccessful publishing

benilovj committed:
Remove AaibReport unsuccessful publish specs

These are already covered in `document_spec`.

benilovj committed:
Add a spec for Document#find

benilovj committed:
Remove redundant #find specs for specialist document types

These cases are all covered by `document_spec`.

benilovj committed:
Extract a shared spec for document subclasses

This shared spec:
- exercises the #save! method for the various subclasses
- validates the Publishing API payload against the JSON schema for the
inidividual document type

benilovj committed:
Move attachment specs into document_spec

The behaviour isn't CmaCase-specific so doesn't belong in that class's spec.

tuzz commented on spec/models/a_document_subclass.rb:
```diff
@@ -0,0 +1,33 @@
+          ]
+        }
+      )
+    end
```
Should we use this method in the `document_spec` too? There's a similar assertion about `First published`.

benilovj replied:
I've refactored `document_spec` to use this method.

tuzz commented on spec/models/document_spec.rb:
```diff
@@ -39,4 +45,216 @@ def self.title
+        }
+      })
+  }
+  let(:document) { MyDocumentType.from_publishing_api(payload) }
```
👍

rgarner commented on spec/models/countryside_stewardship_grant_spec.rb:
```diff
@@ -1,93 +1,7 @@
-    end
-  end
+  let(:payload) { Payloads.countryside_stewardship_grant_content_item }
+  it_behaves_like "a Document subclass"
```
How is a `Document` subclass expected to behave? I went and looked. Effectively, it looks like we mean "it saves only documents valid against the schema". Could we say that here instead?

benilovj replied:
I've renamed the shared examples to something with a bit more meaning.

rgarner commented on spec/models/document_spec.rb:
```diff
@@ -39,4 +45,216 @@ def self.title
+
+  let(:finder_schema) {
+    {
+      base_path: "/my-docyment-types",
```
typo

benilovj replied:
fixed.

tuzz commented on spec/spec_helper.rb:
```diff
@@ -63,4 +63,10 @@
+    document.delete("updated_at")
+    document.delete("publication_state")
+    document
+  end
```
This is also in `document_spec`

benilovj replied:
I've gotten rid of the duplication.

tuzz commented on spec/support/payloads/document.rb:
```diff
@@ -303,7 +303,7 @@ def self.drug_safety_update_content_item(attrs = {})
   end
 
-  def self.employment_tribunal_decision_content_item(attrs)
+  def self.employment_tribunal_decision_content_item(attrs = {})
```
What do you think about making these methods child factories of `parent`?

benilovj replied:
Yes, but could we defer work on the factories to another PR? I did the minimum to keep things working here but would like to revisit the test factories wholesale on another PR.

tuzz replied:
👍

rgarner commented on spec/models/document_spec.rb:
```diff
@@ -39,4 +45,216 @@ def self.title
+      stub_any_publishing_api_patch_links
+
+      c = MyDocumentType.find(payload["content_id"])
+      expect(c.save!).to eq(true)
```
Is there a `context` in which it wouldn't? Both `save!` and `publish!` come with a bang, which suggests either mutable state (core Ruby style) or a variant of a method that would raise an error (ActiveRecord style, qv `find` vs `find!`, `save` vs `save!`). If it's neither of these styles, should we lose the `!`?

EDIT: I had a little look in both `Document#save!` and `Document#publish!`. It appears the first is "noisy" – will raise a `RecordNotSaved` error if anything goes wrong, and the second is wrapped in a handler block that returns `true` or `false`. So I guess we could just say `c.save!` here and drop the expectation. `publish!`, on the other hand, probably needs to be stripped of its bang or made to behave consistently :)

benilovj replied:
I agree – `save!` doesn't behave correctly when there's a network issue. I have a fix for it locally but I'll raise a separate pull request because the fix requires a change to the production code, whereas I'd prefer to keep this as a test-only change.

benilovj committed:
Use helper `saved_for_the_first_time` method in document_spec

This commit reduces duplication, and groups some methods in a dedicated
helper module.
